# syntax-transform/3.1/0001-space-before-argument-parentheses.patch
#
# This file is part of mkxp.
#
# Copyright (C) 2013 - 2021 Amaryllis Kulla <ancurio@mapleshrine.eu>
#
# mkxp is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# mkxp is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with mkxp.  If not, see <http://www.gnu.org/licenses/>.

# In Ruby 1.8, when calling a function with more than one argument, it is
# allowed to have a space between the function name and the `(` but shows the
# warning "don't put space before argument parentheses". In Ruby versions
# newer than 1.8, this syntax causes the expression after the function name to
# be interpreted as a single argument instead. As a consequence, this code
# snippet prints `12`, `22`, `12` and `22` in Ruby 1.8 but `22`, `22`, `22` and
# `22` in Ruby versions newer than 1.8.
#
#   def f(x)
#     x * 2
#   end
#   p f (1) + 10
#   p(f (1) + 10)
#   a = f (1) + 10
#   p a
#   a = (f (1) + 10)
#   p a
#
# Also, this code snippet works in Ruby 1.8 but results in a syntax error in
# Ruby versions newer than 1.8 because the `(1, 2)` gets interpreted as a single
# argument.
#
#   def f(x, y)
#     p x
#     p y
#   end
#   f (1, 2)

--- a/parse.y
+++ b/parse.y
@@ -1249,6 +1249,7 @@ static int looking_at_eol_p(struct parser_params *p);
 %type <node> p_args p_args_head p_args_tail p_args_post p_arg
 %type <node> p_value p_primitive p_variable p_var_ref p_expr_ref p_const
 %type <node> p_kwargs p_kwarg p_kw
+%type <node> mkxp_legacy_expr mkxp_legacy_call_args2 mkxp_open_args
 %type <id>   keyword_variable user_variable sym operation operation2 operation3
 %type <id>   cname fname op f_rest_arg f_block_arg opt_f_block_arg f_norm_arg f_bad_arg
 %type <id>   f_kwrest f_label f_arg_asgn call_op call_op2 reswords relop dot_or_colon
@@ -1311,6 +1312,7 @@ static int looking_at_eol_p(struct parser_params *p);
 %token tSTRING_END	"terminator"
 %token tSTRING_DEND	"'}'"
 %token tSTRING_DBEG tSTRING_DVAR tLAMBEG tLABEL_END
+%token tMKXP_LEGACY_LPAREN_ARG	"( arg"
 
 /*
  *	precedence table
@@ -1369,6 +1371,77 @@ program		:  {
 		    }
 		;
 
+mkxp_legacy_expr: command_call
+                | mkxp_legacy_expr keyword_and mkxp_legacy_expr
+                    {
+                        $$ = logop(p, idAND, $1, $3, &@2, &@$);
+                    }
+                | mkxp_legacy_expr keyword_or mkxp_legacy_expr
+                    {
+                        $$ = logop(p, idOR, $1, $3, &@2, &@$);
+                    }
+                | keyword_not mkxp_legacy_expr
+                    {
+                        $$ = call_uni_op(p, method_cond(p, $2, &@2), METHOD_NOT, &@1, &@$);
+                    }
+                | '!' command_call
+                    {
+                        $$ = call_uni_op(p, method_cond(p, $2, &@2), '!', &@1, &@$);
+                    }
+                | arg
+                ;
+
+mkxp_legacy_call_args2
+                : arg_value ',' args opt_block_arg
+                    {
+                    /*%%%*/
+                        $$ = arg_blk_pass(list_concat(NEW_LIST($1, &@1), $3), $4);
+                    /*% %*/
+                    /*% ripper: args_add_block!(args_add!(args_add!(args_new!, $1), $3), $4) %*/
+                    }
+                | assocs opt_block_arg
+                    {
+                    /*%%%*/
+                        $$ = $1 ? NEW_LIST(new_hash(p, $1, &@1), &@1) : 0;
+                        $$ = arg_blk_pass($$, $2);
+                    /*% %*/
+                    /*% ripper: args_add_block!(args_add!(args_new!, bare_assoc_hash!($1)), $2) %*/
+                    }
+                | arg_value ',' assocs opt_block_arg
+                    {
+                    /*%%%*/
+                        $$ = NEW_LIST($1, &@1);
+                        $$ = $3 ? list_concat($$, NEW_LIST(new_hash(p, $3, &@3), &@3)) : $$;
+                        $$ = arg_blk_pass($$, $4);
+                    /*% %*/
+                    /*% ripper: args_add_block!(args_add!(args_add!(args_new!, $1), bare_assoc_hash!($3)), $4) %*/
+                    }
+                | arg_value ',' args ',' assocs opt_block_arg
+                    {
+                    /*%%%*/
+                        $$ = list_concat(NEW_LIST($1, &@1), $3);
+                        $$ = $5 ? arg_append(p, $$, new_hash(p, $5, &@5), &@$) : $$;
+                        $$ = arg_blk_pass($$, $6);
+                    /*% %*/
+                    /*% ripper: args_add_block!(args_add!(args_add!(args_add!(args_new!, $1), $3), bare_assoc_hash!($5)), $6) %*/
+                    }
+                | block_arg
+                    /*% ripper[brace]: args_add_block!(args_new!, $1) %*/
+                ;
+
+mkxp_open_args  : call_args
+                | tMKXP_LEGACY_LPAREN_ARG {SET_LEX_STATE(EXPR_ENDARG);} ')'
+                    {
+                        rb_warn0("don't put space before argument parentheses");
+                        $$ = 0;
+                    }
+                | tMKXP_LEGACY_LPAREN_ARG mkxp_legacy_call_args2 {SET_LEX_STATE(EXPR_ENDARG);} ')'
+                    {
+                        rb_warn0("don't put space before argument parentheses");
+                        $$ = $2;
+                    }
+                ;
+
 top_compstmt	: top_stmts opt_terms
 		    {
 			$$ = void_stmts(p, $1);
@@ -2845,14 +2918,14 @@ command_args	:   {
 			 */
 			int lookahead = 0;
 			switch (yychar) {
-			  case '(': case tLPAREN: case tLPAREN_ARG: case '[': case tLBRACK:
+			  case '(': case tLPAREN: case tLPAREN_ARG: case tMKXP_LEGACY_LPAREN_ARG: case '[': case tLBRACK:
 			    lookahead = 1;
 			}
 			if (lookahead) CMDARG_POP();
 			CMDARG_PUSH(1);
 			if (lookahead) CMDARG_PUSH(0);
 		    }
-		  call_args
+		  mkxp_open_args
 		    {
 			/* call_args can be followed by tLBRACE_ARG (that does CMDARG_PUSH(0) in the lexer)
 			 * but the push must be done after CMDARG_POP() in the parser.
@@ -3009,6 +3082,14 @@ primary		: literal
 		    /*% %*/
 		    /*% ripper: paren!($2) %*/
 		    }
+		| tMKXP_LEGACY_LPAREN_ARG mkxp_legacy_expr {SET_LEX_STATE(EXPR_ENDARG);} rparen
+		    {
+		    /*%%%*/
+		        if (nd_type_p($2, NODE_SELF)) $2->nd_state = 0;
+		        $$ = $2;
+		    /*% %*/
+		    /*% ripper: paren!($2) %*/
+		    }
 		| tLPAREN compstmt ')'
 		    {
 		    /*%%%*/
@@ -9871,12 +9952,18 @@ parser_yylex(struct parser_params *p)
 	return '~';
 
       case '(':
+	if (mkxp_is_syntax_transform_active(p, 1, 8)) {
+	    p->command_start = TRUE;
+	}
 	if (IS_BEG()) {
 	    c = tLPAREN;
 	}
 	else if (!space_seen) {
 	    /* foo( ... ) => method call, no ambiguity */
 	}
+	else if (mkxp_is_syntax_transform_active(p, 1, 8) && IS_lex_state(EXPR_ARG)) {
+	    rb_warn0("don't put space before argument parentheses");
+	}
 	else if (IS_ARG() || IS_lex_state_all(EXPR_END|EXPR_LABEL)) {
 	    c = tLPAREN_ARG;
 	}
@@ -9888,7 +9975,7 @@ parser_yylex(struct parser_params *p)
 	COND_PUSH(0);
 	CMDARG_PUSH(0);
 	SET_LEX_STATE(EXPR_BEG|EXPR_LABEL);
-	return c;
+	return mkxp_is_syntax_transform_active(p, 1, 8) && c == tLPAREN_ARG ? tMKXP_LEGACY_LPAREN_ARG : c;
 
       case '[':
 	p->lex.paren_nest++;
